Support modifying segmentInfos.counter in IndexWriter

guojialiang92 opened the pull request:
### Description

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

This PR aims to address issue [14362](https://github.com/apache/lucene/issues/14362). This issue includes a discussion of the benefits of this modification.

### Tests

- simple scenario
  - I added test `TestIndexWriter#testAdvanceSegmentInfosCounter`, the writer index a bunch of docs, then advance its counter.
- crash and recovery scenario
  - I added test `TestIndexWriter#testAdvanceSegmentCounterInCrashAndRecoveryScenario`, the writer index a bunch of docs, then close it, start a new writer on the same index, and advance its counter.
- Concurrent writing and modification scenario
  - Modified the test `ThreadedIndexingAndSearchingTestCase`, in the concurrent write thread, increased the logic of randomly modifying the segment counter, and checked after all write threads ended.


### Checklist
- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/lucene/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have given Lucene maintainers [access](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the main branch.
- [x] I have run ./gradlew check.
- [x] I have added tests for my changes.

guojialiang92 committed:
IndexWriter support advance segmentInfos counter

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

guojialiang92 committed:
fix UT

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:
```diff
@@ -5037,4 +5037,45 @@ public void testDocValuesSkippingIndexWithoutDocValues() throws Exception {
+      writer.commit();
+    }
+
+    assertEquals(1041, writer.getSegmentInfosCounter());
```
I think it's sufficient to assert that the counter value is `>=1000` i.e. higher than what you advanced it to. `SegmentInfos.counter` is public and gets incremented each time a segment is named, including background merges. Asserting on an exact value can make the test flaky.

guojialiang92 replied:
You're right, I have already made modifications.

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -1427,6 +1427,24 @@ public synchronized void advanceSegmentInfosVersion(long newVersion) {
+  }
+
+  /** Returns the {@link SegmentInfos#counter}. */
+  public synchronized long getSegmentInfosCounter() {
```
Do we really need this synchronized?

guojialiang92 replied:
I have removed the synchronization to avoid affecting performance when it is heavily called in some scenarios.

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -1427,6 +1427,24 @@ public synchronized void advanceSegmentInfosVersion(long newVersion) {
+   *
+   * @lucene.internal
+   */
+  public synchronized void advanceSegmentInfosCounter(long newCounter) {
```
Do we need to call `changed()` or increment change count here?

guojialiang92 replied:
From the `IndexWriter#changed` interface definition, it should indeed be called.

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:
```diff
@@ -5037,4 +5037,47 @@ public void testDocValuesSkippingIndexWithoutDocValues() throws Exception {
+    }
+    long beforeAdvanceSegmentCounter = writer.getSegmentInfosCounter();
+    writer.advanceSegmentInfosCounter(1);
+    assertEquals(beforeAdvanceSegmentCounter, writer.getSegmentInfosCounter());
```
Let's add a similar `gte` check here as in the assert below, instead of assuming that counter value is 1.

guojialiang92 replied:
I have modified.

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:
```diff
@@ -5037,4 +5037,47 @@ public void testDocValuesSkippingIndexWithoutDocValues() throws Exception {
+    assertEquals(beforeAdvanceSegmentCounter, writer.getSegmentInfosCounter());
+
+    writer.advanceSegmentInfosCounter(1000);
+    assertEquals(1000, writer.getSegmentInfosCounter());
```
Do we need this check? Let's assert with `gte` to avoid flakiness.

guojialiang92 replied:
I have modified.

guojialiang92 committed:
Some CR-related modifications

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

guojialiang92 commented:
Thanks for helping with the code review, I have made modifications according to the suggestions

vigyasharma commented:
I think we can add a couple more tests to make it robust.
1. Some tests around concurrency –  index with multiple threads, then advance the counter in one of the threads, and validate behavior. You can look at `ThreadedIndexingAndSearchingTestCase` and its derived tests for motivation.
2. A test for the crash-recovery scenario, which I suppose it the primary use case. We could make the writer index a bunch of docs, then kill it, start a new writer on the same index, and advance its counter.

vigyasharma commented:
Also, IIUC `IndexWriter#advanceSegmentInfosVersion()` was added to handle similar scenarios for NRT replication (Lucene's native segment replication implementation). I'm curious why we didn't run into the need to advance `SegmentInfos#counter` at that time. Do you remember, @mikemccand (I know it's been a while! (: )?

guojialiang92 committed:
add test

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

guojialiang92 committed:
add test

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

guojialiang92 commented:
Thanks, @vigyasharma 
> I think we can add a couple more tests to make it robust.
> 
> 1. Some tests around concurrency –  index with multiple threads, then advance the counter in one of the threads, and validate behavior. You can look at `ThreadedIndexingAndSearchingTestCase` and its derived tests for motivation.
> 2. A test for the crash-recovery scenario, which I suppose it the primary use case. We could make the writer index a bunch of docs, then kill it, start a new writer on the same index, and advance its counter.

I have added the following tests according to the suggestions: 
1. For crash-recovery scenario, I have added `TestIndexWriter#testAdvanceSegmentCounterInCrashAndRecoveryScenario`.
2. For scenarios with multi-threaded concurrent writes, I have modified the test `ThreadedIndexingAndSearchingTestCase`, in the concurrent write thread, increased the logic of randomly modifying the segment counter, and checked after all write threads ended.

guojialiang92 commented:
Thanks, @vigyasharma 
I also looked at Lucene's native segment replication, just sharing my personal opinion.

> Also, IIUC `IndexWriter#advanceSegmentInfosVersion()` was added to handle similar scenarios for NRT replication (Lucene's native segment replication implementation). I'm curious why we didn't run into the need to advance `SegmentInfos#counter` at that time. Do you remember, @mikemccand (I know it's been a while! (: )?

In the code comments of Lucene's native segment replication, the risk of file conflicts is also mentioned, but no additional processing is done. From a robustness perspective, perhaps control should also be carried out. The relevant code is as follows：
ReplicaNode#fileIsIdentical (**Segment name was reused!  This is rare but possible and otherwise devastating**)
```
  private boolean fileIsIdentical(String fileName, FileMetaData srcMetaData) throws IOException {

    FileMetaData destMetaData = readLocalFileMetaData(fileName);
    if (destMetaData == null) {
      // Something went wrong in reading the file (it's corrupt, truncated, does not exist, etc.):
      return false;
    }

    if (Arrays.equals(destMetaData.header(), srcMetaData.header()) == false
        || Arrays.equals(destMetaData.footer(), srcMetaData.footer()) == false) {
      // Segment name was reused!  This is rare but possible and otherwise devastating:
      if (isVerboseFiles()) {
        message("file " + fileName + ": will copy [header/footer is different]");
      }
      return false;
    } else {
      return true;
    }
  }
```

guojialiang92 commented:
Hi, @vigyasharma 
Could you please help me take a look again? Do you have any other suggestions?

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -1427,6 +1427,25 @@ public synchronized void advanceSegmentInfosVersion(long newVersion) {
+
+  /** Returns the {@link SegmentInfos#counter}. */
+  public long getSegmentInfosCounter() {
+    this.ensureOpen();
```
Do we care that the writer is open for this API?

guojialiang92 replied:
Yeah, we currently cannot obtain `IndexWriter#segmentInfos` or `IndexWriter#segmentInfos.counter` from `IndexWriter` and Lucene users cannot perform `IndexWriter#segmentInfos.counter` related verification.

vigyasharma reviewed (commented):
Apologies for the delay, these changes look great, @guojialiang92. I love the new tests, and how you're testing multi-threaded scenarios. 

I have some minor comments. Please also add a `CHANGES.txt` entry as well and I'll merge this.

guojialiang92 committed:
Merge remote-tracking branch 'origin/main' into dev/indexwriter_support_advance_segment_counter

guojialiang92 committed:
Modify tests and CHANGES.txt

Signed-off-by: guojialiang <guojialiang.2012@bytedance.com>

guojialiang92 commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/ThreadedIndexingAndSearchingTestCase.java:
```diff
@@ -189,6 +190,19 @@ public void run() {
                   }
 
+                  // Maybe advance segment counter
+                  if (random().nextBoolean()) {
```
Thanks, I have modified the test and CHANGES.txt.

guojialiang92 commented:
Thanks, @vigyasharma 
I have made the modifications as suggested, please reivew the code again.

vigyasharma committed:
Move changes entry to 10.3
